waku/v2/protocol/waku_filter/waku_filter.nim

```diff
@@ -188,43 +189,72 @@ method init*(wf: WakuFilter) =
   let rng = crypto.newRng()
   var wf = WakuFilter(rng: rng,
                       peerManager: peerManager, 
-                      pushHandler: handler)
+                      pushHandler: handler,
+                      timeout: timeout)
   wf.init()
-  
   return wf
 
 proc setPeer*(wf: WakuFilter, peer: RemotePeerInfo) =
   wf.peerManager.addPeer(peer, WakuFilterCodec)
   waku_filter_peers.inc()
 
+#clear the failed peer table if subscriber was able to connect.
+proc handleClientSuccess*(wf: WakuFilter, subscribers: seq[Subscriber]){.raises: [Defect, KeyError], async.} = 
+  for subscriber in subscribers:
+    var subKey: string = $(subscriber)
+    if wf.failedPeers.hasKey(subKey):
+      wf.failedPeers.del(subKey)
+
+# If we have already failed to send message to this peer,
+# check for elapsed time and if it's been too long, remove the peer.
+proc handleClientError*(wf: WakuFilter, subscribers: seq[Subscriber]){.raises: [Defect, KeyError], async.} = 
+  for subscriber in subscribers:
+    var subKey: string = $(subscriber)
+    if wf.failedPeers.hasKey(subKey):
+      var elapsedTime = Moment.now() - wf.failedPeers[subKey]
+      if(elapsedTime > wf.timeout):
+        trace "Remove peer if timeout has reached for", peer=subscriber
+        var index = wf.subscribers.find(subscriber)
+        wf.subscribers.delete(index)
+        wf.failedPeers.del(subKey)
+    else:
+      # add the peer to the failed peers table.
+      wf.failedPeers[subKey] = Moment.now() 
+  return
 
 proc handleMessage*(wf: WakuFilter, topic: string, msg: WakuMessage) {.async.} =
   # Handle WakuMessage according to filter protocol
   trace "handle message in WakuFilter", topic=topic, msg=msg
-
+  var handleMessageFailed = false
+  var failedSubscriber: seq[Subscriber]
+  var connectedSubscribers: seq[Subscriber]
   for subscriber in wf.subscribers:
     if subscriber.filter.pubSubTopic != "" and subscriber.filter.pubSubTopic != topic:
       trace "Subscriber's filter pubsubTopic does not match message topic", filter=subscriber.filter.pubSubTopic, topic=topic
       continue
-
+    
     for filter in subscriber.filter.contentFilters:
       if msg.contentTopic == filter.contentTopic:
         trace "Found matching contentTopic", filter=filter, msg=msg
         let push = FilterRPC(requestId: subscriber.requestId, push: MessagePush(messages: @[msg]))
-        
         let connOpt = await wf.peerManager.dialPeer(subscriber.peer, WakuFilterCodec)
-
         if connOpt.isSome:
           await connOpt.get().writeLP(push.encode().buffer)
+          connectedSubscribers.add(subscriber)
         else:
           # @TODO more sophisticated error handling here
+          handleMessageFailed = true
+          failedSubscriber.add(subscriber)
           error "failed to push messages to remote peer"
           waku_filter_errors.inc(labelValues = [dialFailure])
         break
+  discard handleClientSuccess(wf, connectedSubscribers)

```

Comment:
Futures (implicit return of `{.async.}`) should not be discarded (see: https://hackmd.io/@vac/BJbgn4QT_). Use `asyncSpawn` instead. However, I'm not sure that this should be a future at all. See my comment elsewhere.

Reply:
resolves in d91eaeb6f12c7deb56d65fb0e275f119d2e56158